### PR TITLE
[TRTLLM-14497][feat] Add BF16/FP8 refit for qwen3.5_397b

### PR DESCRIPTION
**Pull request by hchings**
## Summary by CodeRabbit

* **New Features**
  * Added support for incremental and partial weight updates for Qwen3.5 models, including vision-language and MoE variants.
  * Supports BF16 and FP8 checkpoint updates, with improved handling of staged projection weights and scaling data.
  * Added safer update-session controls to finalize or discard incomplete updates.

* **Bug Fixes**
  * Improved reliability for GPU kernels when tensor memory alignment is unsuitable.
  * Corrected weight loading behavior for supplied model configurations and multimodal checkpoints.

* **Tests**
  * Added coverage for partial loading and multi-GPU weight updates.

This change enables incremental BF16 weight refit for the Qwen3.5-397B rollout model.
  NeMo-RL transfers a model update through multiple size-bounded buckets, so related Gated
  DeltaNet projections_Q/K/V/Z and B/A_may arrive in separate reload() calls. Qwen3.5 must
  retain incomplete projection groups across calls, fuse them only when complete, and
  validate the complete update before post-load transformations run. The implementation
  introduces explicit begin/finalize/abort lifecycle boundaries and preserves in-place
  parameter storage required by CUDA Graph replay.

  The model-specific behavior is isolated to Qwen3_5MoeHfWeightMapper. The MLPerf 397B
  model is currently the only model requiring this cross-bucket staging because of its
  Qwen3.5 Gated DeltaNet projection layout and BF16 MoE representation. Other weight
  mappers inherit no-op lifecycle methods, so their successful loading and refit behavior
  is unchanged. The generic RLHF worker owns the lifecycle calls because only the transport
  layer knows the first bucket, finalization boundary, and failure boundary; placing a
  397B-specific type check there would couple generic update infrastructure to one model.

  The change also fixes a TP8 alignment issue for a small GDN tensor view that can begin on
  a 16-byte boundary on odd TP shards, while the CuTe kernel requires 32-byte alignment.
  Aligned configurations retain the existing zero-copy path.

  Validation covers reduced-depth, production-width Qwen3.5-397B BF16 refit with warm CUDA-
  graph execution and post-refit Hugging Face reference-logit comparison:

  - TP4 one-shot refit
  - TP4 incremental/partial refit
  - Two-node TP8 one-shot refit (Not committed due to multi-node)
  - Two-node TP8 incremental/partial refit (Not committed due to multi-node)

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [ ] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/models/checkpoints/hf/qwen3_5_weight_mapper.py [hunk 1]
```diff
@@ -555,10 +667,15 @@ def _remap_dense_mlp_weights(self, weights: dict) -> dict:
...
 
-    def preprocess_weights(self, weights: dict) -> dict:
+    def preprocess_weights(self,
+                           weights: dict,
+                           allow_partial_loading: bool = False) -> dict:
         quant_algo = self.config.quant_config.quant_algo
 
         normalized_weights = self._normalize_weight_names(weights)
+        if allow_partial_loading:
+            normalized_weights = self._stage_partial_split_projections(
```
Comment (QiJune) [inline tensorrt_llm/_torch/models/checkpoints/hf/qwen3_5_weight_mapper.py:673]: Could you please confirm the behavior for ModelOpt FP8 checkpoints whose split projection scales arrive as `.weight_scale` rather than `.weight_scale_inv`?

Here staging runs before `_normalize_scale_names()`, while `_stage_partial_split_projections()` couples block-FP8 weights only with `.weight_scale_inv`. It looks like `.weight_scale` tensors could be emitted separately and renamed afterward, leaving the corresponding weights pending. Similarly, per-tensor FP8 weights may reach requantization before a later scale bucket arrives.

The unit test calls the staging helper directly with already-normalized `.weight_scale_inv` keys, so it does not exercise this public `preprocess_weights()` ordering. If only checkpoints that natively use `.weight_scale_inv` are in scope, this may be fine; otherwise, could we add a preprocess-level multi-bucket test to confirm?

**Review by QiJune — APPROVED** [review]

File: tensorrt_llm/llmapi/rlhf_utils.py [hunk 2]
```diff
@@ -152,6 +157,9 @@ def update_weights(self, ipc_handles: Optional[dict] = None):
                 torch.cuda.empty_cache()
 
         except Exception as e:
+            self.engine.model_engine.model_loader.abort_update_weights()
```
Comment (yechank-nvidia) [inline tensorrt_llm/llmapi/rlhf_utils.py:205]: If an incremental update fails after earlier buckets have already been copied into live parameters, `abort_update_weights()` only clears the mapper-owned staging state.

For example, a subset of the parameters may have been updated while the remaining parameters retain their previous values, leaving the model in a mixed or inconsistent checkpoint state.

Can you confirm whether the RL orchestrator guarantees that the worker cannot resume serving until a complete update succeeds, or that the worker is discarded after such a failure?

**Review by yechank-nvidia — APPROVED** [review]

File: tensorrt_llm/_torch/modules/fla/fused_sigmoid_gating_recurrent.py [hunk 3]
```diff
@@ -278,6 +278,10 @@ def _flashinfer_gdn_decode(
     assert T_per_seq == 1, (
         f"_flashinfer_gdn_decode expects standard decode (T_per_seq == 1), got "
         f"{T_per_seq}; _can_use_flashinfer_gdn_decode should keep T == N")
+    # TP8 leaves eight BF16 ``a`` values per Qwen3.5 GDN shard. Odd shards can
+    # start 16 bytes into fused projection storage, but CuTe requires 32B.
+    if a_bat.data_ptr() % 32:
+        a_bat = a_bat.clone(memory_format=torch.contiguous_format)
```
Comment (shuyixiong) [inline tensorrt_llm/_torch/modules/fla/fused_sigmoid_gating_recurrent.py:302]: Could you confirm whether this operation works with CUDA Graphs?

File: tensorrt_llm/llmapi/rlhf_utils.py (hunk 2 above)
Comment (hchings) [reply tensorrt_llm/llmapi/rlhf_utils.py:205]: When update_weights fails, NeMo-RL's GRPO orchestrator raises a RuntimeError immediately (see [grpo.py L2315](https://github.com/NVIDIA-NeMo/RL/blob/main/nemo_rl/algorithms/grpo.py#L2315)), aborting the whole training job before any further rollout is dispatched. So the worker never resumes rollout in the stale mixed-weight state.

File: tensorrt_llm/_torch/models/checkpoints/hf/qwen3_5_weight_mapper.py (hunk 1 above)
Comment (hchings) [reply tensorrt_llm/_torch/models/checkpoints/hf/qwen3_5_weight_mapper.py:673]: Thanks for pointing this out. Only the native block-FP8 checkpoints of `qwen3.5-397b/35b` using `weight_scale_inv` are in scope for this MR, as so far MLPerf work (which this MR is for) hasn't used modelopt ckpt. Right now modelopt FP8 ckpt will fail directly during refit.
 
I'll add support of modelopt FP8 ckpt for refit in the future if there's a solid use case.

File: tensorrt_llm/_torch/modules/fla/fused_sigmoid_gating_recurrent.py (hunk 3 above)
Comment (hchings) [reply tensorrt_llm/_torch/modules/fla/fused_sigmoid_gating_recurrent.py:302]: Yes, I added a new test (TP8 BF16 ckpt w/ Cuda graph enabled) in this [commit](https://github.com/NVIDIA/TensorRT-LLM/pull/16642/changes/434f2715089b5323f1796d1bac27c6c5d81dce15#diff-fbd8ad5cfc57f6fdb238eab1e3d76dac52579013c1b6388a61a1b4101c8a1cd9) to validate this, and it passes w/ both 35B and 397B on 2 X 4-GPUs GB200 locally, which log showing that this clone line was invoked during test.  In CI, I put this new test in 8-GPU B200 so as to test TP8. 

The clone is safe because it executes during graph capture. Its output comes from CUDA Graph_s private memory pool, so replay reuses the same address and only replays the captured copy and CuTe kernels. It is temporary activation storage; refitted parameters still retain their original addresses.

**Comment by hchings** [comment]
/bot run --stage-list "DGX_B200-8_GPUs-PyTorch-Ray-1, DGX_B200-4_GPUs-PyTorch-Ray-1"

File: tensorrt_llm/_torch/modules/fla/fused_sigmoid_gating_recurrent.py (hunk 3 above)
Comment (shuyixiong) [reply tensorrt_llm/_torch/modules/fla/fused_sigmoid_gating_recurrent.py:302]: Thanks for the clarification.

File: tests/unittest/_torch/ray_orchestrator/multi_gpu/test_llm_update_weights_multi_gpu.py [hunk 4]
```diff
@@ -129,82 +182,523 @@ def test_llm_update_weights_fp8(model_dir, fp8_model_dir):
...
 
-        # Generate filter_list from model weight keys by removing layer prefix
-        # e.g., "model.layers.41.input_layernorm.weight" -> "input_layernorm.weight"
-        layer_prefix_pattern = re.compile(r"^model\.layers\.\d+\.")
-        filter_set = set()
-        for name, _ in hf_model.all_weights[hf_model.device_id]:
-            suffix = layer_prefix_pattern.sub("", name)
-            filter_set.add(suffix)
-        filter_list = list(filter_set)
+@pytest.mark.high_cuda_memory
```
Comment (shuyixiong) [inline tests/unittest/_torch/ray_orchestrator/multi_gpu/test_llm_update_weights_multi_gpu.py:644]: The new tests in this file will not be picked up automatically by CI, so please add them manually to test_db.

File: tests/unittest/_torch/ray_orchestrator/multi_gpu/test_llm_update_weights_multi_gpu.py [hunk 5]
```diff
@@ -129,82 +182,523 @@ def test_llm_update_weights_fp8(model_dir, fp8_model_dir):
...
-        # Finalize the update weights
-        llm._collective_rpc("update_weights", (None,))
+    _attach_to_tp8_ray_cluster(monkeypatch)
+    _run_qwen35_35b_tp8_multinode_update()
 
-        llm_logits, ref_logits = run_generate(llm, hf_model, prompts, sampling_params)
-        compare_logits(llm_logits, ref_logits)
 
-    del hf_model
+@pytest.mark.part0
```
Comment (shuyixiong) [inline tests/unittest/_torch/ray_orchestrator/multi_gpu/test_llm_update_weights_multi_gpu.py]: Could you set up a dedicated part for the Qwen3.5 refit? Preferably, it should not reuse the existing part0 and part1, as too many update weight tests within a single test session can lead to instability .

Comment (hchings) [reply tests/unittest/_torch/ray_orchestrator/multi_gpu/test_llm_update_weights_multi_gpu.py]: Updated as part5 in 723193665812ea3470b65cbbfc9fe94c158af333.

File: tests/unittest/_torch/ray_orchestrator/multi_gpu/test_llm_update_weights_multi_gpu.py (hunk 4 above)
Comment (hchings) [reply tests/unittest/_torch/ray_orchestrator/multi_gpu/test_llm_update_weights_multi_gpu.py:644]: Updated in 723193665812ea3470b65cbbfc9fe94c158af333.

**Review by dpitman-nvda — APPROVED** [review]
Approval for infra

**Review by BowenFu — APPROVED** [review]
Reviewed the current head. The incremental refit lifecycle and Qwen3.5 BF16/FP8 checkpoint paths look sound for the targeted MLPerf workflow.
